reana_server/utils.py

```diff
@@ -130,10 +132,57 @@ def remove_fetched_workflows_dir(tmpdir: str) -> None:
         shutil.rmtree(tmpdir)
 
 
-def mv_workflow_files(source: str, target: str) -> None:
-    """Move files from one directory to another."""
-    for entry in os.listdir(source):
-        shutil.move(os.path.join(source, entry), target)
+def mv_workflow_files(source: str, target: str, reana_spec: Dict) -> None:
+    """Move workflow's input files from one directory to another.
+
+    :param source: Directory from which the files will be moved.
+    :param target: Directory where the files will be moved to.
+    :param reana_spec: REANA specification used to decide which files to move.
+    """
+    # FIXME: use `is_relative_to` from the standard library when moving to Python 3.9
+    def is_relative_to(path: pathlib.Path, base: pathlib.Path) -> bool:
+        """Check whether `path` is contained inside `base`."""
+        try:
+            path.relative_to(base)
+            return True
+        except ValueError:
+            return False
+
+    def validate_paths(paths_str: Sequence[str]) -> List[pathlib.Path]:
+        """Check whether the list of paths is valid or not."""
+        paths = [pathlib.Path(path) for path in paths_str]
+
+        unique_paths = set()
+        for path in paths:
+            if path.is_absolute():
+                raise REANAValidationError(f"Path cannot be absolute: {path}")
+            if not path.parts:
+                raise REANAValidationError("Path cannot be empty")
+            if ".." in path.parts:
+                raise REANAValidationError(f"Path cannot contain '..': {path}")
+            if path in unique_paths:
+                raise REANAValidationError(f"Path declared multiple times: {path}")
+            unique_paths.add(path)
+
+        for x, y in itertools.permutations(paths, r=2):
+            if is_relative_to(x, y):
+                raise REANAValidationError(f"Invalid paths: '{y}' is a prefix of '{x}'")
+
+        return paths
```

Comment:
I think we should validate the paths present in `inputs.directories` and `inputs.files` even when using reana-client. Should I open an issue to move this logic to reana-commons and to use it in both reana-client and reana-server?